### ISSUE TITLE
Crash in CLion with 1.3.1 rendering simple link

### ISSUE THREAD

**Issue by tamboril** [vsch/idea-multimarkdown#109]
After switching to JavaFX and trying another page, it crashed again.  

I can't seem to format this link so that it doesn't cause the error:

    [ABC](doc/news.md)

Here is the stack:

    String index out of range: -4
    java.lang.StringIndexOutOfBoundsException: String index out of range: -4
	at java.lang.String.substring(String.java:1919)
	at kotlin.text.StringsKt__StringsJVMKt.substring(StringsJVM.kt:136)
	at kotlin.text.StringsKt.substring(Unknown Source)
	at com.vladsch.idea.multimarkdown.util.GitHubLinkMatcher.computeMatchText(GitHubLinkMatcher.kt:222)
	at com.vladsch.idea.multimarkdown.util.GitHubLinkMatcher$isOnlyCompletionMatchValid$2.invoke(GitHubLinkMatcher.kt:96)
	at com.vladsch.idea.multimarkdown.util.GitHubLinkMatcher$isOnlyCompletionMatchValid$2.invoke(GitHubLinkMatcher.kt:29)
	at kotlin.SynchronizedLazyImpl.getValue(Lazy.kt:125)
	at kotlin.LazyKt.getValue(Lazy.kt:77)
	at com.vladsch.idea.multimarkdown.util.GitHubLinkMatcher.isOnlyCompletionMatchValid(GitHubLinkMatcher.kt)
	at com.vladsch.idea.multimarkdown.util.GitHubLinkResolver.getMatchedRefs(GitHubLinkResolver.kt:215)
	at com.vladsch.idea.multimarkdown.util.GitHubLinkResolver.resolve(GitHubLinkResolver.kt:154)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownLinkRenderer.getLinkTarget(MultiMarkdownLinkRenderer.java:78)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownLinkRenderer.renderLink(MultiMarkdownLinkRenderer.java:98)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownLinkRenderer.render(MultiMarkdownLinkRenderer.java:124)
	at org.pegdown.ToHtmlSerializer.visit(ToHtmlSerializer.java:215)
	at org.pegdown.ast.ExpLinkNode.accept(ExpLinkNode.java:35)
	at org.pegdown.ToHtmlSerializer.visitChildrenSkipFirst(ToHtmlSerializer.java:606)
	at org.pegdown.ToHtmlSerializer.visitChildren(ToHtmlSerializer.java:594)
	at org.pegdown.ToHtmlSerializer.visit(ToHtmlSerializer.java:579)
	at org.pegdown.ast.SuperNode.accept(SuperNode.java:43)
	at org.pegdown.ToHtmlSerializer.visitChildrenSkipFirst(ToHtmlSerializer.java:606)
	at org.pegdown.ToHtmlSerializer.visitChildren(ToHtmlSerializer.java:594)
	at org.pegdown.ToHtmlSerializer.printTag(ToHtmlSerializer.java:620)
	at org.pegdown.ToHtmlSerializer.printBreakBeforeTag(ToHtmlSerializer.java:627)
	at org.pegdown.ToHtmlSerializer.visit(ToHtmlSerializer.java:315)
	at org.pegdown.ast.ParaNode.accept(ParaNode.java:35)
	at org.pegdown.ToHtmlSerializer.visitChildrenSkipFirst(ToHtmlSerializer.java:606)
	at org.pegdown.ToHtmlSerializer.visitChildren(ToHtmlSerializer.java:594)
	at org.pegdown.ToHtmlSerializer.visit(ToHtmlSerializer.java:120)
	at org.pegdown.ast.RootNode.accept(RootNode.java:76)
	at org.pegdown.ToHtmlSerializer.toHtml(ToHtmlSerializer.java:90)
	at com.vladsch.idea.multimarkdown.editor.resources.JavaFxHtmlGenerator.toHtml(JavaFxHtmlGenerator.kt:54)
	at com.vladsch.idea.multimarkdown.editor.MarkdownPreviewFileEditor$HtmlUpdater.run(MarkdownPreviewFileEditor.kt:266)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:238)

**Comment by vsch** [comment]
@tamboril, I made a build with tracing to the console. It has happened to another user but I cannot see under what configuration or bugs it can happen. It means that the file containing the link is not under its vcs root directory nor under the project root since its path is shorter than vcs root or project root.

<https://github.com/vsch/idea-multimarkdown/raw/master/dist/idea-multimarkdown.1.3.1.3.zip>

You will need to launch CLion from the command line so all log messages are visible, not only idea-multimarkdown. I really want to figure out what configuration or bug is causing this. I cannot duplicate this on my system.

**Comment by tamboril** [comment]
In this case, the file is above the project root, but at the top of the Git repository.
Same behavior, though, if I open a non-git project in a totally different directory, then try to open the readme.md file from the other project.

**Comment by tamboril** [comment]
Sorry didn't mean to close

**Comment by vsch** [comment]
@tamboril thank you for the clarification. I never expected files to be outside the project root or have vcs roots above the project root. In this case I the project root takes precedence because the plugin does not scan for vcs roots above project directory.

I'll create myself a test project with similar configuration and finally resolve this issue.

Thank you very much for your time and effort. I'll post a quick fix as soon as I have it done.

**Comment by tamboril** [comment]
The file above the project root is there because there are cpp, java, ..., etc., which are project roots for different JetBrains IDE's.  Everything is in one repo, though; so there has to be a readme.md at the top (so it shows up automatically in GitHub or Visual Studio Online), and that is necessarily above the language- and IDE-specific subdirectories.

I ran it with a [trace](https://github.com/vsch/idea-multimarkdown/files/72734/t.zip), which is attached.  Thank you for your attention to this.

**Comment by vsch** [comment]
@tamboril, thank you for taking the time to help me solve this. I had an inverted model of project/repo relationship. It will be fixed shortly.

**Comment by vsch** [comment]
@tamboril, the fix is in <https://github.com/vsch/idea-multimarkdown/raw/master/dist/idea-multimarkdown.1.3.1.4.zip>

Vcs roots now searched above project directory.

**Comment by tamboril** [comment]
Vladimir, now it renders OK, but the relative link is unresolved.  The link is to the doc directory, which is not in the project, either.  Like `[ABC](./doc/news.md)`

**Comment by vsch** [comment]
@tamboril, 

Link resolution requires that the files being accessed are in the project tree in one form or another because the plugin uses the file based indices of the IDE to find targets. The directories with targets for links should be as part of the project in one form or another. Otherwise, they will not be found..

I setup vcs-root in the top directory and two projects without their own vcs roots in subdirectories. Could you give me a screenshot of the project tree showing the source file (the one with the link) and the target of the link? if this is not proprietary information. I want to make sure I have the test project setup similar to what you have.

<img width="1133" alt="screen shot 2015-12-27 at 3 06 20 pm" src="https://cloud.githubusercontent.com/assets/10299086/12011795/69c733a2-acab-11e5-9ba0-5cd6e06419f3.png">

**Comment by tamboril** [comment]
Neither file is in the project.  I'm just trying to make a main page with a link to a page in `doc/`.  This would be for non-language-specific documentation of my project.  Eventually, I may (or may not) have some content underneath the CLion project root, but that should not be required to link two documents outside the project.

A screenshot would be useless, because non-project files don't show up in the tree.  But I can still open them, and should be able to navigate links among them, like I do at GitHub.

    gitroot/
        readme.md
        doc/
            news.md
        cpp/
            src/
        java/
            src/

**Comment by vsch** [comment]
The plugin used to scan for its own files independent of the IDE index mechanisms but it was causing all kinds of integration problems. 

Now it uses what the IDE provides. If the file is not in the project in some form so the IDE includes it in an index then the plugin will not see it either. The plugin has to use what the IDE provides especially when it comes to file searches. Some projects have over 100,000 files and scanning for these or writing my own indexing mechanism which won't play well with the rest of the application is not an option. 

You can type the link in any form and preview all files but the plugin won't be of help in navigation or error checking unless the files are part of a project.

**Comment by tamboril** [comment]
I understand.  I guess I need to get another tool to view/edit my  documentation, then.  It doesn't make sense to me for my documents to be down with my C++ source when many of the documents are generic to both cpp and java implementations below them, just so they can be seen by multimarkdown.

**Comment by vsch** [comment]
@tamboril you don't need to put them in the project directory. You just need to add a module that contains these documents to the project. I haven't worked with CLion but I am sure it is the same as the other IDEs from JetBrains, the modules can come from anywhere. 

It is necessary that the files be in some form part of the workspace so that the IDE can find them, and hence the plugin. 

Now for viewing you can open these files and view them and edit them. As far as getting help with links and navigation, then they need to be part of some project element. 

You can also try GFM plugin which uses GitHub for rendering and uses a Chromium based browser for accurate GitHub rendering, Markdown Support from JetBrains, the split screen I use was based on that plugin, there is also the original idea-markdown, but it has been the least updated of all available plugins.

Thank you very much for taking the time in helping me resolve the issues and I am sorry I was not able to resolve all of them.

Have a Happy New Year!

**Comment by tamboril** [comment]
Vladimir, thanks to your hints, I got it to work.  While CLion doesn't have an explicit modules function like Idea--it's based on CMake files--there is a "Change Project Root" function, which I used to move the project root up a level.  The `.idea` directory stays where it was, but references `..` in the main xml file.  Then it sees the .md files and correctly resolves the links.

**Comment by vsch** [comment]
I am glad you got it working. From my experience with JetBrains products they are extremely flexible and powerful but do require understanding and adopting their working model. I only started working with them a year ago and was very frustrated because they didn't fit my understanding. 

Everything seemed to take more configuration than necessary and I kept getting lost in all the dialogs and options. Once I understood their approach it became clear why it is there and how to use it. Now, I couldn't work without them and it will take a few more years to learn a significant part of the features they offer. They are a universe of functionality created by a team of extremely bright individuals. It is a fun trip if you enjoy learning and discovery.
